miner/algo_greedy_buckets.go

```diff
@@ -0,0 +1,228 @@
+package miner
+
+import (
+	"crypto/ecdsa"
+	"math/big"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// / To use it:
+// / 1. Copy relevant data from the worker
+// / 2. Call buildBlock
+// / 2. If new bundles, txs arrive, call buildBlock again
+// / This struct lifecycle is tied to 1 block-building task
+type greedyBucketsBuilder struct {
+	inputEnvironment *environment
+	chainData        chainData
+	builderKey       *ecdsa.PrivateKey
+	interrupt        *int32
+	gasUsedMap       map[*types.TxWithMinerFee]uint64
+	algoConf         algorithmConfig
+}
+
+func newGreedyBucketsBuilder(
+	chain *core.BlockChain, chainConfig *params.ChainConfig, algoConf *algorithmConfig,
+	blacklist map[common.Address]struct{}, env *environment, key *ecdsa.PrivateKey, interrupt *int32,
+) *greedyBucketsBuilder {
+	if algoConf == nil {
+		algoConf = &algorithmConfig{
+			EnforceProfit:          true,
+			ExpectedProfit:         nil,
+			ProfitThresholdPercent: defaultProfitThreshold,
+		}
+	}
+	return &greedyBucketsBuilder{
+		inputEnvironment: env,
+		chainData:        chainData{chainConfig: chainConfig, chain: chain, blacklist: blacklist},
+		builderKey:       key,
+		interrupt:        interrupt,
+		gasUsedMap:       make(map[*types.TxWithMinerFee]uint64),
+		algoConf:         *algoConf,
+	}
+}
+
+func (b *greedyBucketsBuilder) commit(envDiff *environmentDiff,
+	transactions []*types.TxWithMinerFee,
+	orders *types.TransactionsByPriceAndNonce,
+	gasUsedMap map[*types.TxWithMinerFee]uint64, retryMap map[*types.TxWithMinerFee]int, retryLimit int,
+) ([]types.SimulatedBundle, []types.UsedSBundle) {
+	var (
+		usedBundles  []types.SimulatedBundle
+		usedSbundles []types.UsedSBundle
+		algoConf     = b.algoConf
+
+		CheckRetryOrderAndReinsert = func(
```

Comment:
I think this kind of retry is likely to be useless.

It will actually never work in the case that make sense - when bundle is inserted successfully but we fail it because it underpays. For it to work we need to update bundle mev gas price and profit with values that bundle got when it failed and reinsert it with a new values to the heap. 

No need to change anything, just an opinion.

Reply:
Updated to only reinsert when error is due to low profit, and added logic to update profit and price prior to reinserting to heap